torch_xla/csrc/runtime/operation_tracker.cc

```diff
@@ -0,0 +1,79 @@
+#include "torch_xla/csrc/runtime/operation_tracker.h"
+
+#include <shared_mutex>
+
+#include "absl/types/span.h"
+#include "torch_xla/csrc/runtime/debug_macros.h"
+#include "torch_xla/csrc/runtime/tf_logging.h"
+
+namespace torch_xla {
+namespace runtime {
+
+OperationTracker::OperationTracker(absl::Span<const std::string> devices) {
+  for (auto& device : devices) {
+    op_counters_.try_emplace(device, device);
+  }
+}
+
+OperationTracker::Operation::Operation(Counter* counter) : counter_(counter) {
+  XLA_CHECK(counter_);
+  counter_->Increment();
+}
+
+OperationTracker::Operation::~Operation() { counter_->Decrement(); }
+
+std::unique_ptr<OperationTracker::Operation> OperationTracker::StartOperation(
+    std::string device) {
+  return std::make_unique<Operation>(&op_counters_.at(device));
+}
+
+void OperationTracker::WaitForDevices(absl::Span<const std::string> devices) {
+  std::vector<std::unique_lock<std::shared_mutex>> locks;
+  locks.reserve(devices.size());
+
+  for (const std::string& device_str : devices) {
+    TF_VLOG(5) << "Blocking new operations on " << device_str;
+    auto lock = op_counters_.at(device_str).BlockNewOperations();
+    locks.emplace_back(std::move(lock));
+
+    TF_VLOG(3) << "Waiting for device execution for " << device_str
+               << " to finish";
+    op_counters_.at(device_str).Wait();
+    TF_VLOG(3) << "Finished operations on device " << device_str;
+  }
+}
+
+void OperationTracker::Counter::Increment() {
+  // Block new operations after Wait() is called. count_ is already atomic, so
+  // atomic so we don't need an exclusive lock to prevent data races.
+  std::shared_lock lock(pending_operations_mu_);
+  auto current = count_.fetch_add(1, std::memory_order_acq_rel) + 1;
+  TF_VLOG(5) << "Incremented operations for " << device_ << " to " << current;
+}
+
+void OperationTracker::Counter::Decrement() {
+  auto current = count_.fetch_sub(1, std::memory_order_acq_rel) - 1;
```

Comment:
not exactly sure about `fetch_sub`'s syntax but I am guessing it returns the value before subtraction?

Reply:
Yes. This is just a quirk of how atomics have to work.

Reply:
After looking at this again, I'm half inclined to replace the atomic counter with locks and a normal int. I don't expect contention to be a concern because there will be relatively few concurrent users of that mutex, and I can stop anyone else from having to read up on `std::atomic` and memory order. What do you think?

Reply:
I suspect most people will ever read into this implementation, both ways works for me.

Reply:
It turns out using a lock+int counter is actually noticeably slower. I'll just take the win instead of interrogating why...